File: examples/15_loop_refinement.py
"""Example 15: Loop Refinement — iterative improvement until quality met.

Scenario: An essay writing pipeline where:
  1. draft: writes an initial draft
  2. review: scores the draft (0-1 quality score)
  3. revise: improves the draft using review feedback
  4. Steps 2-3 repeat until quality >= 0.8 or max 5 iterations

This demonstrates the Loop modifier — neograph's support for cyclical
graphs. The loop compiles to a LangGraph conditional back-edge.

Patterns demonstrated:
  1. Self-loop: single node refines its own output
  2. Multi-node loop body as sub-construct with Loop
  3. Loop inside a sub-construct (parent sees clean I/O)
  4. ForwardConstruct with self.loop() — explicit cycle primitive
  5. Produce + validate with Loop (input != output)

Run:
    python examples/15_loop_refinement.py
"""

from __future__ import annotations

from pydantic import BaseModel

from neograph import (
    compile,
    construct_from_functions,
    node,
    run,
)
from neograph import Loop


# -- Schemas ------------------------------------------------------------------

class Draft(BaseModel, frozen=True):
    content: str
    iteration: int = 0
    score: float = 0.0


class ReviewResult(BaseModel, frozen=True):
    score: float
    feedback: str


class Essay(BaseModel, frozen=True):
    content: str
    final_score: float
    iterations: int


# =============================================================================
# Pattern 1: Self-loop — single node refines its own output
# =============================================================================
#
# Graph: start -> seed -> refine -(loop)-> end
#
# refine takes a Draft, improves it, returns a Draft. Loops until
# score >= 0.8 or max 5 iterations.

def demo_self_loop():
    """Self-loop: single node refines its own output."""
    call_count = [0]

    @node(outputs=Draft)
    def seed() -> Draft:
        return Draft(content="Initial draft about microservices", iteration=0, score=0.0)

    @node(
        outputs=Draft,
        loop_when=lambda d: d is None or d.score < 0.8,
        max_iterations=5,
    )
    def refine(draft: Draft) -> Draft:
        """Each iteration improves the score by 0.3."""
        call_count[0] += 1
        new_score = draft.score + 0.3
        return Draft(
            content=f"v{call_count[0]}: refined draft",
            iteration=draft.iteration + 1,
            score=new_score,
        )

    pipeline = construct_from_functions("self-loop", [seed, refine])
    graph = compile(pipeline)
    result = run(graph, input={"node_id": "self-loop-demo"})

    # Loop result is an append-list: all iterations preserved
    history = result["refine"]
    final = history[-1]

    print("=== Pattern 1: Self-loop ===")
    print(f"Iterations: {len(history)}")
    for i, draft in enumerate(history):
        print(f"  [{i+1}] score={draft.score:.1f} content={draft.content!r}")
    print(f"Final: score={final.score:.1f}, iteration={final.iteration}")
    print()


# =============================================================================
# Pattern 2: Multi-node loop body as sub-construct
# =============================================================================
#
# Graph: start -> draft -> [refine: review -> revise] -(loop)-> finalize -> end
#
# The review+revise cycle is a sub-construct that takes Draft in, produces
# Draft out. The Loop modifier on the sub-construct makes it repeat.

def demo_multi_node_loop():
    """Multi-node loop: review + revise cycle as a looping sub-construct."""
    review_count = [0]

    @node(outputs=Draft)
    def draft() -> Draft:
        return Draft(content="Initial essay about authentication", score=0.0)

    @node(outputs=ReviewResult)
    def review(draft: Draft) -> ReviewResult:
        review_count[0] += 1
        score = 0.3 * review_count[0]
        return ReviewResult(
            score=min(score, 1.0),
            feedback=f"Iteration {review_count[0]}: {'needs work' if score < 0.8 else 'approved'}",
        )

    @node(outputs=Draft)
    def revise(draft: Draft, review: ReviewResult) -> Draft:
        return Draft(
            content=f"Revised: {review.feedback}",
            iteration=draft.iteration + 1,
            score=review.score,
        )

    @node(outputs=Essay)
    def finalize(refine: Draft) -> Essay:
        return Essay(content=refine.content, final_score=refine.score, iterations=refine.iteration)

    # Loop body = sub-construct with Draft in, Draft out
    refine = construct_from_functions(
        "refine", [review, revise], input=Draft, output=Draft,
    ) | Loop(when=lambda d: d is None or d.score < 0.8, max_iterations=10)

    pipeline = construct_from_functions("writer", [draft, refine, finalize])
    graph = compile(pipeline)
    result = run(graph, input={"node_id": "multi-loop-demo"})

    # refine field is an append-list from the loop
    history = result["refine"]
    essay = result["finalize"]

    print("=== Pattern 2: Multi-node loop (sub-construct) ===")
    print(f"Review iterations: {review_count[0]}")
    for i, d in enumerate(history):
        print(f"  [{i+1}] score={d.score:.1f} content={d.content!r}")
    print(f"Final essay: score={essay.final_score}, iterations={essay.iterations}")
    print()


# =============================================================================
# Pattern 3: Loop inside a sub-construct (hidden from parent)
# =============================================================================
#
# Graph: start -> seed -> [refine: write -> improve -(loop)->] -> finalize -> end
#
# The refinement loop is internal to the sub-construct. The parent pipeline
# sees only Draft -> Essay.

def demo_loop_in_sub_construct():
    """Loop inside a sub-construct — parent sees clean I/O."""
    improve_count = [0]

    @node(outputs=Draft)
    def write(topic: Draft) -> Draft:
        return Draft(content="First draft", score=0.5)

    @node(
        outputs=Draft,
        loop_when=lambda d: d is None or d.score < 0.8,
        max_iterations=5,
    )
    def improve(write: Draft) -> Draft:
        improve_count[0] += 1
        return Draft(
            content=f"Improved v{improve_count[0]}",
            iteration=write.iteration + 1,
            score=write.score + 0.2,
        )

    refine_sub = construct_from_functions(
        "refine", [write, improve],
        input=Draft, output=Draft,
    )

    @node(outputs=Draft)
    def seed() -> Draft:
        return Draft(content="topic: distributed systems", score=0.0)

    @node(outputs=Essay)
    def finalize(refine: Draft) -> Essay:
        return Essay(content=refine.content, final_score=refine.score, iterations=refine.iteration)

    pipeline = construct_from_functions("writer", [seed, refine_sub, finalize])
    graph = compile(pipeline)
    result = run(graph, input={"node_id": "sub-loop-demo"})

    essay = result["finalize"]

    print("=== Pattern 3: Loop inside sub-construct ===")
    print(f"Internal improve iterations: {improve_count[0]}")
    print(f"Final essay: score={essay.final_score}, iterations={essay.iterations}")
    # Sub-construct internals don't leak to parent result
    print(f"Parent sees 'refine' output: {'refine' in result}")
    print(f"Parent does NOT see 'write': {'write' not in result}")
    print(f"Parent does NOT see 'improve': {'improve' not in result}")
    print()


# =============================================================================
# Pattern 4: ForwardConstruct with self.loop()
# =============================================================================
#
# Graph: start -> draft -> [loop-body: review -> revise] -(loop)-> end
#
# self.loop() is the explicit cycle primitive for ForwardConstruct. Python
# for/while loops in forward() trace the body once but don't produce graph
# cycles (same limitation as torch.jit.trace). self.loop() compiles to a
# sub-construct with Loop modifier — a real back-edge in the graph.

def demo_forward_loop():
    """ForwardConstruct with self.loop() — explicit cycle primitive."""
    from neograph import ForwardConstruct, Node, register_scripted

    review_count = [0]

    register_scripted(
        "fc_draft_ex15",
        lambda _in, _cfg: Draft(content="Initial essay about distributed systems", score=0.0),
    )

    def fc_review(_in, _cfg):
        review_count[0] += 1
        score = 0.3 * review_count[0]
        return ReviewResult(
            score=min(score, 1.0),
            feedback=f"Iteration {review_count[0]}: {'needs work' if score < 0.8 else 'approved'}",
        )

    register_scripted("fc_review_ex15", fc_review)

    def fc_revise(_in, _cfg):
        return Draft(
            content=f"Revised v{review_count[0]}",
            iteration=review_count[0],
            score=0.3 * review_count[0],
        )

    register_scripted("fc_revise_ex15", fc_revise)

    class Writer(ForwardConstruct):
        draft  = Node.scripted("draft", fn="fc_draft_ex15", outputs=Draft)
        review = Node.scripted("review", fn="fc_review_ex15", outputs=ReviewResult)
        revise = Node.scripted("revise", fn="fc_revise_ex15", outputs=Draft)

        def forward(self, topic):
            d = self.draft(topic)
            d = self.loop(
                body=[self.review, self.revise],
                when=lambda r: r is None or r.score < 0.8,
                max_iterations=10,
            )(d)
            return d

    graph = compile(Writer())
    result = run(graph, input={"node_id": "forward-loop-demo"})

    print("=== Pattern 4: ForwardConstruct with self.loop() ===")
    print(f"Review iterations: {review_count[0]}")
    print(f"Draft output: {result.get('draft')}")
    # The loop body is a sub-construct; its output appears under the
    # sub-construct name, not individual node names.
    for key, val in result.items():
        if isinstance(val, list) and val and hasattr(val[0], "score"):
            for i, d in enumerate(val):
                print(f"  [{i+1}] score={d.score:.1f} content={d.content!r}")
    print()


# =============================================================================
# Pattern 5: Produce + Validate with Loop (input != output)
# =============================================================================
#
# Graph: start -> seed -> [produce_validate: produce -> validate] -(loop)-> report -> end
#
# The sub-construct takes ProduceInput and returns Validated — different types.
# The producer fails on the first call (simulating a transient error) and
# succeeds on the second. The validator checks for errors. The Loop condition
# retries until validation passes or max iterations are exhausted.

def demo_produce_validate():
    """Produce + validate with Loop — input != output, retry on failure."""

    class ProduceInput(BaseModel, frozen=True):
        request_id: str
        payload: str

    class ProduceOutput(BaseModel, frozen=True):
        request_id: str
        data: str
        error: str | None = None

    class Validated(BaseModel, frozen=True):
        request_id: str
        data: str
        errors: list[str]

    class Report(BaseModel, frozen=True):
        request_id: str
        data: str
        attempts: int

    produce_calls = [0]

    @node(outputs=ProduceOutput)
    def produce(topic: ProduceInput) -> ProduceOutput:
        """Fails on first call, succeeds on second."""
        produce_calls[0] += 1
        if produce_calls[0] == 1:
            return ProduceOutput(
                request_id=topic.request_id,
                data="",
                error="transient failure: service unavailable",
            )
        return ProduceOutput(
            request_id=topic.request_id,
            data=f"generated data for '{topic.payload}'",
            error=None,
        )

    @node(outputs=Validated)
    def check(produce: ProduceOutput) -> Validated:
        """Check the produced output for errors."""
        errors: list[str] = []
        if produce.error:
            errors.append(produce.error)
        if not produce.data:
            errors.append("empty data")
        return Validated(
            request_id=produce.request_id,
            data=produce.data,
            errors=errors,
        )

    # Sub-construct: ProduceInput -> Validated (different types)
    produce_validate = construct_from_functions(
        "produce_validate", [produce, check],
        input=ProduceInput, output=Validated,
    ) | Loop(
        when=lambda v: v is None or bool(v.errors),
        max_iterations=5,
    )

    @node(outputs=ProduceInput)
    def seed() -> ProduceInput:
        return ProduceInput(request_id="req-42", payload="quarterly report")

    @node(outputs=Report)
    def report(produce_validate: Validated) -> Report:
        return Report(
            request_id=produce_validate.request_id,
            data=produce_validate.data,
            attempts=produce_calls[0],
        )

    pipeline = construct_from_functions(
        "produce-pipeline", [seed, produce_validate, report],
    )
    graph = compile(pipeline)
    result = run(graph, input={"node_id": "produce-validate-demo"})

    history = result["produce_validate"]
    final_report = result["report"]

    print("\n" + "=" * 60)
    print("PRODUCE+VALIDATE LOOP (input != output)")
    print("=" * 60)
    print(f"Producer calls: {produce_calls[0]}")
    for i, v in enumerate(history):
        status = "PASS" if not v.errors else f"FAIL ({', '.join(v.errors)})"
        print(f"  [{i+1}] {status} data={v.data!r}")
    print(f"Final report: request_id={final_report.request_id}, "
          f"data={final_report.data!r}, attempts={final_report.attempts}")
    print()


# =============================================================================

if __name__ == "__main__":
    demo_self_loop()
    demo_multi_node_loop()
    demo_loop_in_sub_construct()
    demo_forward_loop()
    demo_produce_validate()

File: docs/design/ts-parity/15_loop_refinement.ts
// TS PARITY SKETCH — hypothetical, NOT compilable/runnable.
// Python source: examples/15_loop_refinement.py
// Target API: docs/design/typescript-port.md (AD-0 transformer form + .pipe() + fluent ForwardConstruct)
//
// This file grounds a feature-parity analysis of neograph's Loop modifier across
// five patterns: self-loop, multi-node loop sub-construct, hidden-loop sub-construct,
// ForwardConstruct self.loop(), and produce+validate (input != output).
//
// Legend for inline notes:
//   // PARITY:DIRECT   — same logic, different syntax
//   // PARITY:REDESIGN — TS limitation forces a different approach
//   // PARITY:BLOCKED  — no clean TS path
//   // PARITY:GAP      — proposed TS API in typescript-port.md doesn't cover this

import { z } from "zod";
import {
  node,
  compile,
  run,
  Loop,
  constructFromFunctions,
  ForwardConstruct,
  Node,
} from "@neograph/core";

// -- Schemas ------------------------------------------------------------------
// PARITY:DIRECT — Pydantic BaseModel(frozen=True) -> Zod object.
// PARITY:GAP — frozen=True has no Zod analogue; z.parse() returns a plain mutable
//   object. Would need Object.freeze() wrapping or a readonly branded type to
//   match the Python immutability guarantee. The example never mutates a model,
//   so behavior is preserved, but the *contract* silently weakens.

const Draft = z.object({
  content: z.string(),
  iteration: z.number().int().default(0),
  score: z.number().default(0.0),
});
type Draft = z.infer<typeof Draft>;

const ReviewResult = z.object({
  score: z.number(),
  feedback: z.string(),
});
type ReviewResult = z.infer<typeof ReviewResult>;

const Essay = z.object({
  content: z.string(),
  final_score: z.number(),
  iterations: z.number().int(),
});
type Essay = z.infer<typeof Essay>;

// =============================================================================
// Pattern 1: Self-loop — single node refines its own output
// =============================================================================

function demoSelfLoop() {
  // PARITY:DIRECT (actually nicer) — Python boxes mutable closure state as
  // `call_count = [0]` because Python closures can't rebind an outer scalar.
  // TS `let` closures rebind natively, so no [0] boxing ceremony.
  let callCount = 0;

  // PARITY:REDESIGN — Python `@node(outputs=Draft)` decorator on a standalone
  // `def seed()` has no TS equivalent (TS decorators only attach to class members).
  // Becomes the node({config}, fn) wrapper. AD-0 transformer reads `(): Draft`
  // return annotation to emit outputs=Draft, so the schema stays implicit.
  // PARITY:GAP — a no-input node forces an empty config object `node({}, ...)`;
  // Python's bare `@node(outputs=Draft)` is cleaner for zero-param seeds.
  const seed = node({}, (): Draft => ({
    content: "Initial draft about microservices",
    iteration: 0,
    score: 0.0,
  }));

  const refine = node(
    {
      // PARITY:DIRECT — loop_when lambda -> loopWhen arrow fn.
      // PARITY:REDESIGN — the `d is None` first-iteration sentinel. Python passes
      //   None; LangGraph.js state defaults are typically `undefined` OR a
      //   default() factory. Which null-ish value the runtime feeds the predicate
      //   on iteration 0 must be pinned, else `d === null` misses an `undefined`.
      //   Mirrors the loop_condition_none_unsafe lint — but TS has TWO empties.
      loopWhen: (d: Draft | null) => d === null || d.score < 0.8,
      maxIterations: 5,
    },
    // AD-0 transformer extracts inputs={draft: Draft}, output=Draft from signature.
    (draft: Draft): Draft => {
      callCount += 1;
      const newScore = draft.score + 0.3;
      return {
        content: `v${callCount}: refined draft`,
        iteration: draft.iteration + 1,
        score: newScore,
      };
    },
  );

  // PARITY:DIRECT — construct_from_functions([...]) passes wrapper handles.
  const pipeline = constructFromFunctions("self-loop", [seed, refine]);
  const graph = compile(pipeline);
  const result = run(graph, { input: { node_id: "self-loop-demo" } });

  // PARITY:GAP — Loop's append-reducer makes result["refine"] a Draft[], but the
  //   NODE output type is Draft. run()'s return is Record<string, unknown>, so
  //   there is no typed accessor that reflects "this one field got list-ified by
  //   a modifier". Python is also untyped at result[...] but the language never
  //   promised more; TS users EXPECT typed results, so this reads as a regression.
  const history = result["refine"] as Draft[];
  // PARITY:DIRECT (minor) — Python `history[-1]` negative index -> length math.
  const final = history[history.length - 1];

  console.log("=== Pattern 1: Self-loop ===");
  console.log(`Iterations: ${history.length}`);
  history.forEach((draft, i) =>
    console.log(`  [${i + 1}] score=${draft.score.toFixed(1)} content=${JSON.stringify(draft.content)}`),
  );
  console.log(`Final: score=${final.score.toFixed(1)}, iteration=${final.iteration}`);
}

// =============================================================================
// Pattern 2: Multi-node loop body as sub-construct
// =============================================================================

function demoMultiNodeLoop() {
  let reviewCount = 0;

  const draft = node({}, (): Draft => ({
    content: "Initial essay about authentication",
    iteration: 0,
    score: 0.0,
  }));

  const review = node({}, (draft: Draft): ReviewResult => {
    reviewCount += 1;
    const score = 0.3 * reviewCount;
    return {
      score: Math.min(score, 1.0),
      feedback: `Iteration ${reviewCount}: ${score < 0.8 ? "needs work" : "approved"}`,
    };
  });

  // PARITY:DIRECT — fan-in: two upstream params (draft, review). AD-0 transformer
  // emits inputs={draft: Draft, review: ReviewResult}; param names ARE the edges.
  const revise = node({}, (draft: Draft, review: ReviewResult): Draft => ({
    content: `Revised: ${review.feedback}`,
    iteration: draft.iteration + 1,
    score: review.score,
  }));

  // PARITY:REDESIGN — the consumer param `refine` is named after the SUB-CONSTRUCT
  //   below, and it's typed Draft (single) while the state field holds Draft[]
  //   (loop append-list). neograph feeds downstream consumers the last element.
  //   The transformer will extract inputs={refine: Draft}; the list->element
  //   unwrap is a runtime convention the TS type system can't see. Fine at
  //   runtime, but `refine: Draft` here vs `Draft[]` in result[] is a type schism.
  const finalize = node({}, (refine: Draft): Essay => ({
    content: refine.content,
    final_score: refine.score,
    iterations: refine.iteration,
  }));

  // PARITY:REDESIGN — the pipe. Python `sub | Loop(...)` -> `.pipe(Loop({...}))`.
  //   Also: input=Draft / output=Draft are Pydantic CLASSES in Python but must be
  //   Zod SCHEMA VALUES here. These are runtime boundary ports the transformer
  //   can't infer from any signature — they're explicit args either way.
  const refine = constructFromFunctions(
    "refine",
    [review, revise],
    { input: Draft, output: Draft },
  ).pipe(
    Loop({ when: (d: Draft | null) => d === null || d.score < 0.8, maxIterations: 10 }),
  );

  const pipeline = constructFromFunctions("writer", [draft, refine, finalize]);
  const graph = compile(pipeline);
  const result = run(graph, { input: { node_id: "multi-loop-demo" } });

  const history = result["refine"] as Draft[];
  const essay = result["finalize"] as Essay;

  console.log("=== Pattern 2: Multi-node loop (sub-construct) ===");
  console.log(`Review iterations: ${reviewCount}`);
  history.forEach((d, i) =>
    console.log(`  [${i + 1}] score=${d.score.toFixed(1)} content=${JSON.stringify(d.content)}`),
  );
  console.log(`Final essay: score=${essay.final_score}, iterations=${essay.iterations}`);
}

// =============================================================================
// Pattern 3: Loop inside a sub-construct (hidden from parent)
// =============================================================================

function demoLoopInSubConstruct() {
  let improveCount = 0;

  // PARITY:REDESIGN — `write(topic: Draft)`: the param `topic` type matches the
  //   sub-construct's input=Draft, so it's a PORT param (reads neo_subgraph_input,
  //   not a peer node). Python resolves this by issubclass(annotation, input).
  //   TS must match extracted typeName "Draft" against the input schema by name/
  //   structure (AD-3 Zod comparison) — issubclass has no TS equivalent.
  const write = node({}, (topic: Draft): Draft => ({
    content: "First draft",
    iteration: 0,
    score: 0.5,
  }));

  const improve = node(
    {
      loopWhen: (d: Draft | null) => d === null || d.score < 0.8,
      maxIterations: 5,
    },
    (write: Draft): Draft => {
      improveCount += 1;
      return {
        content: `Improved v${improveCount}`,
        iteration: write.iteration + 1,
        score: write.score + 0.2,
      };
    },
  );

  const refineSub = constructFromFunctions(
    "refine",
    [write, improve],
    { input: Draft, output: Draft },
  );

  const seed = node({}, (): Draft => ({
    content: "topic: distributed systems",
    iteration: 0,
    score: 0.0,
  }));

  const finalize = node({}, (refine: Draft): Essay => ({
    content: refine.content,
    final_score: refine.score,
    iterations: refine.iteration,
  }));

  const pipeline = constructFromFunctions("writer", [seed, refineSub, finalize]);
  const graph = compile(pipeline);
  const result = run(graph, { input: { node_id: "sub-loop-demo" } });

  const essay = result["finalize"] as Essay;

  console.log("=== Pattern 3: Loop inside sub-construct ===");
  console.log(`Internal improve iterations: ${improveCount}`);
  console.log(`Final essay: score=${essay.final_score}, iterations=${essay.iterations}`);
  // PARITY:DIRECT — state-hygiene assertions (sub-construct internals don't leak).
  console.log(`Parent sees 'refine' output: ${"refine" in result}`);
  console.log(`Parent does NOT see 'write': ${!("write" in result)}`);
  console.log(`Parent does NOT see 'improve': ${!("improve" in result)}`);
}

// =============================================================================
// Pattern 4: ForwardConstruct with self.loop() — explicit cycle primitive
// =============================================================================

function demoForwardLoop() {
  let reviewCount = 0;

  // PARITY:DIRECT — raw scripted fns with (state, config) signature.
  const fcDraft = (_in: unknown, _cfg: unknown): Draft => ({
    content: "Initial essay about distributed systems",
    iteration: 0,
    score: 0.0,
  });
  const fcReview = (_in: unknown, _cfg: unknown): ReviewResult => {
    reviewCount += 1;
    const score = 0.3 * reviewCount;
    return {
      score: Math.min(score, 1.0),
      feedback: `Iteration ${reviewCount}: ${score < 0.8 ? "needs work" : "approved"}`,
    };
  };
  const fcRevise = (_in: unknown, _cfg: unknown): Draft => ({
    content: `Revised v${reviewCount}`,
    iteration: reviewCount,
    score: 0.3 * reviewCount,
  });

  // PARITY:REDESIGN — ForwardConstruct subclass. Node.scripted(...) as class
  //   FIELDS ports directly (class members CAN take TS decorators/initializers).
  //   forward() body is symbolically traced: `this.draft(topic)` returns a JS
  //   Proxy, recording the call. Python uses _Proxy.__getattr__; TS uses a
  //   Proxy get-trap. DIRECT per matrix.
  class Writer extends ForwardConstruct {
    draft = Node.scripted("draft", { fn: "fc_draft_ex15", outputs: Draft });
    review = Node.scripted("review", { fn: "fc_review_ex15", outputs: ReviewResult });
    revise = Node.scripted("revise", { fn: "fc_revise_ex15", outputs: Draft });

    forward(topic: any) {
      let d = this.draft(topic);
      // PARITY:REDESIGN — self.loop() is the EXPLICIT cycle primitive. This
      //   example NEVER uses `if (proxy)` branching, so it dodges the hardest
      //   ForwardConstruct port problem entirely: Python re-traces via
      //   `_Proxy.__bool__`, but JS `if (proxy)` always truthy and JS cannot
      //   intercept boolean coercion (no __bool__). Branch-heavy examples would
      //   need the .gt()/.then()/.else() fluent API; THIS one only needs .loop(),
      //   which is a plain method call — so the loop primitive is a clean port.
      d = this.loop({
        body: [this.review, this.revise],
        when: (r: ReviewResult | null) => r === null || r.score < 0.8,
        maxIterations: 10,
      })(d);
      return d;
    }
  }

  // PARITY:DIRECT — string-keyed scripted registry passed to compile.
  //   Python `scripted={...}` dict -> TS `{ scripted: {...} }` options object.
  //   The fn="fc_draft_ex15" string indirection ports 1:1 (module-level Map).
  const graph = compile(new Writer(), {
    scripted: {
      fc_draft_ex15: fcDraft,
      fc_review_ex15: fcReview,
      fc_revise_ex15: fcRevise,
    },
  });
  const result = run(graph, { input: { node_id: "forward-loop-demo" } });

  console.log("=== Pattern 4: ForwardConstruct with self.loop() ===");
  console.log(`Review iterations: ${reviewCount}`);
  console.log(`Draft output: ${JSON.stringify(result["draft"])}`);
  // PARITY:REDESIGN — Python duck-types the loop output via hasattr(val[0],
  //   "score") over result.values(). TS erases types at runtime, so "does this
  //   list element have a .score" needs a runtime shape check or the Zod schema.
  for (const val of Object.values(result)) {
    if (Array.isArray(val) && val.length && typeof (val[0] as any)?.score === "number") {
      (val as Draft[]).forEach((d, i) =>
        console.log(`  [${i + 1}] score=${d.score.toFixed(1)} content=${JSON.stringify(d.content)}`),
      );
    }
  }
}

// =============================================================================
// Pattern 5: Produce + Validate with Loop (input != output)
// =============================================================================

function demoProduceValidate() {
  // PARITY:DIRECT — nested schemas defined in function scope.
  const ProduceInput = z.object({ request_id: z.string(), payload: z.string() });
  type ProduceInput = z.infer<typeof ProduceInput>;

  // PARITY:DIRECT — `error: str | None = None` -> nullable+default.
  const ProduceOutput = z.object({
    request_id: z.string(),
    data: z.string(),
    error: z.string().nullable().default(null),
  });
  type ProduceOutput = z.infer<typeof ProduceOutput>;

  const Validated = z.object({
    request_id: z.string(),
    data: z.string(),
    errors: z.array(z.string()),
  });
  type Validated = z.infer<typeof Validated>;

  const Report = z.object({
    request_id: z.string(),
    data: z.string(),
    attempts: z.number().int(),
  });
  type Report = z.infer<typeof Report>;

  let produceCalls = 0;

  // PARITY:REDESIGN — `topic: ProduceInput` is a PORT param (type matches
  //   input=ProduceInput). Name+type identity match across the transformer/runtime
  //   seam replaces Python issubclass.
  const produce = node({}, (topic: ProduceInput): ProduceOutput => {
    produceCalls += 1;
    if (produceCalls === 1) {
      return { request_id: topic.request_id, data: "", error: "transient failure: service unavailable" };
    }
    return { request_id: topic.request_id, data: `generated data for '${topic.payload}'`, error: null };
  });

  const check = node({}, (produce: ProduceOutput): Validated => {
    const errors: string[] = [];
    if (produce.error) errors.push(produce.error);
    if (!produce.data) errors.push("empty data");
    return { request_id: produce.request_id, data: produce.data, errors };
  });

  // PARITY:REDESIGN — sub-construct with input != output (ProduceInput -> Validated).
  //   `.pipe(Loop(...))` on a boundary-ported sub-construct. The when predicate
  //   reads bool(v.errors); TS `v.errors.length > 0` (no truthy-list coercion).
  const produceValidate = constructFromFunctions(
    "produce_validate",
    [produce, check],
    { input: ProduceInput, output: Validated },
  ).pipe(
    Loop({
      when: (v: Validated | null) => v === null || v.errors.length > 0,
      maxIterations: 5,
    }),
  );

  const seed = node({}, (): ProduceInput => ({ request_id: "req-42", payload: "quarterly report" }));

  const report = node({}, (produce_validate: Validated): Report => ({
    request_id: produce_validate.request_id,
    data: produce_validate.data,
    attempts: produceCalls,
  }));

  const pipeline = constructFromFunctions("produce-pipeline", [seed, produceValidate, report]);
  const graph = compile(pipeline);
  const result = run(graph, { input: { node_id: "produce-validate-demo" } });

  const history = result["produce_validate"] as Validated[];
  const finalReport = result["report"] as Report;

  console.log("PRODUCE+VALIDATE LOOP (input != output)");
  console.log(`Producer calls: ${produceCalls}`);
  history.forEach((v, i) => {
    const status = v.errors.length === 0 ? "PASS" : `FAIL (${v.errors.join(", ")})`;
    console.log(`  [${i + 1}] ${status} data=${JSON.stringify(v.data)}`);
  });
  console.log(
    `Final report: request_id=${finalReport.request_id}, data=${JSON.stringify(finalReport.data)}, attempts=${finalReport.attempts}`,
  );
}

// PARITY:DIRECT — top-level runner.
demoSelfLoop();
demoMultiNodeLoop();
demoLoopInSubConstruct();
demoForwardLoop();
demoProduceValidate();
